feat: Coordinator 구현

nil-ian opened the pull request:
❗️ Closes #1 Coordinator 구현


### 📝 작업 유형

- 신규 기능 추가

### 📙 작업 내역

- Coordinator 구현
  - CorodinatorType 프로토콜 정의
  - Home, VIdeoUpload, Profile 코디네이터 구현
  - SceneFactory 구현
  - TabBarCoordinator 구현
  - AppCoordinator 구현 및 SceneDelegate와 연결

<br>
<br>

nil-ian committed:
feat: #1 - Coordinator 구현

CorodinatorType 프로토콜 정의

nil-ian committed:
feat: #1 - Coordinator 구현

SceneDelegate 연결을 위한 Tuist 스크립트 수정

nil-ian committed:
feat: #1 - Coordinator 구현

Home, VIdeoUpload, Profile 코디네이터 구현

nil-ian committed:
feat: #1 - SceneFactory 구현

Coordinator에서 ViewController를 생성하는 보일러 플레이트 코드 반복 제거를 위한 팩토리 객체 구현

nil-ian committed:
feat: #1 - TabBarCoordinator 구현

SceneFactory를 생성자 주입받아 활용하여 불필요한 코드 간소화

nil-ian committed:
refactor: #1 - SceneFactory 적용

SceneFactor를 활용하여 기존에 작업해놓은 코드에서 반복되는 코드 제거

nil-ian committed:
feat: #1 - Coordinator 연결

AppCoordinator 구현 및 SceneDelegate와 연결

Hansangjin98 commented on Targets/Feature/Sources/Common/Classes/SceneFactory.swift:
```diff
@@ -0,0 +1,47 @@
+//
+//  SceneFactory.swift
+//  PhoChak
+//
```
모듈명 Feature로 바꿔야겠다!

Hansangjin98 commented on Targets/Feature/Sources/Common/Protocols/CoordinatorType.swift:
```diff
@@ -0,0 +1,36 @@
+//
+//  CoordinatorType.swift
+//  Service
+//
```
여기도 모듈명~!

Hansangjin98 commented on Targets/Feature/Sources/Home/HomeCoordinator.swift:
```diff
@@ -0,0 +1,49 @@
+    self.sceneFactory = dependency.sceneFactory
+  }
+
+  // MARK: Methods
+  func start() {
+    router.pushViewController(sceneFactory.create(scene: .home), animated: true)
+  }
+
+  func startPush() -> UINavigationController {
+    .init(rootViewController: sceneFactory.create(scene: .home))
+  }
```
HomeCoordinator를 포함해서 탭바에 들어가는 각각의 코디네이터들은 start 메서드를 사용하지 않는데 차라리 구현부를 비워두면 어떨까?

그저 개인적인 생각이긴 한데 startPush 메서드도 navigationController만 생성해서 반환하는 용도면 더 적절한 네이밍이 있을 것 같아!

nil-ian replied:
오 좋은생각!
우리가 코디네이터에서 화면 전환하는 로직을 수행하는 메소드를 `start` 라는 이름을 붙인것 처럼 관례를 따라서 메소드 네이밍을 했는데  통일성을 위한건지 위와 같은 상황에도 `startPush` 라는 이름을 쓰더라고~

나도 더 좋은 직관적인 네이밍이 있을것 같아서 한번 고민해보고 좋은 네이밍이 있다면 바꾸는게 좋을것 같다!

Hansangjin98 replied:
아아 그렇구나!!
직관적인 네이밍 한번 같이 생각해보자~!

Hansangjin98 commented on Targets/Feature/Sources/Common/Classes/SceneFactory.swift:
```diff
@@ -0,0 +1,47 @@
+  case profile
+}
+
+protocol SceneFactoryType {
+  func create(scene: Scene) -> UIViewController
+}
+
+public final class SceneFactory: SceneFactoryType {
```
질문) 모든 코디네이터에서 인터페이스가 아닌 구현체 자체를 주입받고 있고, 주입받은 이후에 자식 코디네이터에게 그대로 다시 주입시키고 있는데 그냥 의존성을 주입해 나가는 것보다 create 메서드를 타입 메서드로 사용하는게 더 깔끔하지 않을까?
팩토리 패턴으로 적용하는 이유가 궁금해!

nil-ian replied:
인터페이스 기반으로 주입해주는게 좋겠다!

- 그대로 다시 주입시키고 있는데 그냥 의존성을 주입해 나가는 것보다 create 메서드를 타입 메서드로 사용하는게 더 깔끔하지 않을까?
  - 나도 `SceneFactory` 내 `create`메소드를 `static` 메소드나 `class` 메소드로 제공할까 고민해봤는데 `SceneFactory` 객체의 핵심 역할을 단순히 말 그대로 Scene을 생성해서 반환해주는 역할이잖아?. 즉 이 객체와 상호작용할 객체는 `Coordinator` 뿐이고 `Coordinator` 들은 서로간에 연결이 가능하니까 `static`을 통해서 정적으로 사용할 필요를 못 느껴서 이렇게 작성을 했어~ 근데 이 부분은 `Feature` 모듈 내에서만 `static` 으로 활용하는게 코드 가독성 측면에서는 훨씬 좋을것 같다~ 의견 부탁쓰!
  
- 팩토리 패턴으로 적용하는 이유가 궁금해!
    - 지금은 각 VC를 생성하기 위한 의존성 객체들이 없으니 코드가 얼마 없어서 팩토리를 사용하지 않아도 괜찮지만 VC를 생성하는 코드(UseCase 생성, Reactor 생성, coordinator delegate 채택 ...etc) 등과 같은 형식적인 코드들이 이전 프로젝트 처럼 각 화면 전환 로직에서 반복될거야. 그 부분에 대해서 개선 필요성을 느꼇고 개선 방안으로 화면을 생성해서 반환하는 역할은 온전히 `SceneFactory` 객체에게 넘김으로써 불필요한 반복되는 코드들을 줄일 수 있어서 적용했으~ 의견 부탁!

Hansangjin98 replied:
오오 그렇구나 많이 배우게 되네!!!
Feature 모듈 외에 다른 곳에서도 적용할 예정이면 그냥 통일성을 생각해서 이대로 가도 괜찮을 것 같기도 하다!

nil-ian replied:
VC를 사용하는건 Feature 모듈 한정이니까 static하게 가는게 코드는 훨씬 깔끔할 것 같다!
이건 조금만 더 고민해볼께~

Hansangjin98 reviewed (commented):
고생 많았어~!!!

nil-ian committed:
feat: #1 - 리뷰 내용 수정

- SceneFactory를 Singleton 타입으로 변경

- startPush 메소드명 createNavWrappedViewController로 변경

nil-ian committed:
feat: #1 - 코디네이터 수정

- 각 코디네이터에서 VC 레퍼런스를 유지하도록 변경

Hansangjin98 reviewed (approved):
굿굿~~~!!
